src/v/storage/segment.h

```diff
@@ -276,8 +278,17 @@ inline bool segment::finished_self_compaction() const {
     return (_flags & bitflags::finished_self_compaction)
            == bitflags::finished_self_compaction;
 }
-inline batch_cache_index& segment::cache() { return *_cache; }
-inline const batch_cache_index& segment::cache() const { return *_cache; }
+inline std::optional<std::reference_wrapper<batch_cache_index>>
```

Comment:
Why not just return pointer? The type models nullable reference which is a bit strange, because reference is supposed to be not nullable.

Reply:
We use optional reference in few places, what I like about this vs a pointer is that it forces caller to check the if the result is there. With pointer you can assume that is not null